<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Data analysis and R programming</title>
  <meta name="description" content="Material for the ‘Data analysis and R programming’ course.">
  <meta name="generator" content="bookdown 0.5 and GitBook 2.6.7">

  <meta property="og:title" content="Data analysis and R programming" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Material for the ‘Data analysis and R programming’ course." />
  <meta name="github-repo" content="lgatto/2017_11_09_Rcourse_Jena" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Data analysis and R programming" />
  
  <meta name="twitter:description" content="Material for the ‘Data analysis and R programming’ course." />
  

<meta name="author" content="Laurent Gatto">


<meta name="date" content="2017-11-04">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="interactive-visualisation.html">
<link rel="next" href="r-programming-concepts-and-tools.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/htmlwidgets-0.9/htmlwidgets.js"></script>
<script src="libs/datatables-binding-0.2/datatables.js"></script>
<link href="libs/dt-core-1.10.12/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.10.12/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.10.12/js/jquery.dataTables.min.js"></script>


<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.1</b> Installation</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#programme"><i class="fa fa-check"></i><b>1.2</b> Programme</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#license-and-credits"><i class="fa fa-check"></i><b>1.3</b> License and credits</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="tidy-data.html"><a href="tidy-data.html"><i class="fa fa-check"></i><b>2</b> Tidy data</a><ul>
<li class="chapter" data-level="2.1" data-path="tidy-data.html"><a href="tidy-data.html#introduction-1"><i class="fa fa-check"></i><b>2.1</b> Introduction</a></li>
<li class="chapter" data-level="2.2" data-path="tidy-data.html"><a href="tidy-data.html#common-spreadsheet-errors"><i class="fa fa-check"></i><b>2.2</b> Common Spreadsheet Errors</a></li>
<li class="chapter" data-level="2.3" data-path="tidy-data.html"><a href="tidy-data.html#key-points"><i class="fa fa-check"></i><b>2.3</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="before-we-start.html"><a href="before-we-start.html"><i class="fa fa-check"></i><b>3</b> Before we start</a><ul>
<li class="chapter" data-level="3.1" data-path="before-we-start.html"><a href="before-we-start.html#what-is-r-what-is-rstudio"><i class="fa fa-check"></i><b>3.1</b> What is R? What is RStudio?</a></li>
<li class="chapter" data-level="3.2" data-path="before-we-start.html"><a href="before-we-start.html#why-learn-r"><i class="fa fa-check"></i><b>3.2</b> Why learn R?</a><ul>
<li class="chapter" data-level="3.2.1" data-path="before-we-start.html"><a href="before-we-start.html#r-does-not-involve-lots-of-pointing-and-clicking-and-thats-a-good-thing"><i class="fa fa-check"></i><b>3.2.1</b> R does not involve lots of pointing and clicking, and that’s a good thing</a></li>
<li class="chapter" data-level="3.2.2" data-path="before-we-start.html"><a href="before-we-start.html#r-code-is-great-for-reproducibility"><i class="fa fa-check"></i><b>3.2.2</b> R code is great for reproducibility</a></li>
<li class="chapter" data-level="3.2.3" data-path="before-we-start.html"><a href="before-we-start.html#r-is-interdisciplinary-and-extensible"><i class="fa fa-check"></i><b>3.2.3</b> R is interdisciplinary and extensible</a></li>
<li class="chapter" data-level="3.2.4" data-path="before-we-start.html"><a href="before-we-start.html#r-works-on-data-of-all-shapes-and-sizes"><i class="fa fa-check"></i><b>3.2.4</b> R works on data of all shapes and sizes</a></li>
<li class="chapter" data-level="3.2.5" data-path="before-we-start.html"><a href="before-we-start.html#r-produces-high-quality-graphics"><i class="fa fa-check"></i><b>3.2.5</b> R produces high-quality graphics</a></li>
<li class="chapter" data-level="3.2.6" data-path="before-we-start.html"><a href="before-we-start.html#r-has-a-large-and-welcoming-community"><i class="fa fa-check"></i><b>3.2.6</b> R has a large and welcoming community</a></li>
<li class="chapter" data-level="3.2.7" data-path="before-we-start.html"><a href="before-we-start.html#not-only-is-r-free-but-it-is-also-open-source-and-cross-platform"><i class="fa fa-check"></i><b>3.2.7</b> Not only is R free, but it is also open-source and cross-platform</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="before-we-start.html"><a href="before-we-start.html#knowing-your-way-around-rstudio"><i class="fa fa-check"></i><b>3.3</b> Knowing your way around RStudio</a></li>
<li class="chapter" data-level="3.4" data-path="before-we-start.html"><a href="before-we-start.html#getting-set-up"><i class="fa fa-check"></i><b>3.4</b> Getting set up</a><ul>
<li class="chapter" data-level="3.4.1" data-path="before-we-start.html"><a href="before-we-start.html#organizing-your-working-directory"><i class="fa fa-check"></i><b>3.4.1</b> Organizing your working directory</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="before-we-start.html"><a href="before-we-start.html#interacting-with-r"><i class="fa fa-check"></i><b>3.5</b> Interacting with R</a></li>
<li class="chapter" data-level="3.6" data-path="before-we-start.html"><a href="before-we-start.html#how-to-learn-more-after-the-workshop"><i class="fa fa-check"></i><b>3.6</b> How to learn more after the workshop?</a></li>
<li class="chapter" data-level="3.7" data-path="before-we-start.html"><a href="before-we-start.html#seeking-help"><i class="fa fa-check"></i><b>3.7</b> Seeking help</a><ul>
<li class="chapter" data-level="3.7.1" data-path="before-we-start.html"><a href="before-we-start.html#use-the-built-in-rstudio-help-interface-to-search-for-more-information-on-r-functions"><i class="fa fa-check"></i><b>3.7.1</b> Use the built-in RStudio help interface to search for more information on R functions</a></li>
<li class="chapter" data-level="3.7.2" data-path="before-we-start.html"><a href="before-we-start.html#i-know-the-name-of-the-function-i-want-to-use-but-im-not-sure-how-to-use-it"><i class="fa fa-check"></i><b>3.7.2</b> I know the name of the function I want to use, but I’m not sure how to use it</a></li>
<li class="chapter" data-level="3.7.3" data-path="before-we-start.html"><a href="before-we-start.html#i-want-to-use-a-function-that-does-x-there-must-be-a-function-for-it-but-i-dont-know-which-one"><i class="fa fa-check"></i><b>3.7.3</b> I want to use a function that does X, there must be a function for it but I don’t know which one…</a></li>
<li class="chapter" data-level="3.7.4" data-path="before-we-start.html"><a href="before-we-start.html#i-am-stuck-i-get-an-error-message-that-i-dont-understand"><i class="fa fa-check"></i><b>3.7.4</b> I am stuck… I get an error message that I don’t understand</a></li>
<li class="chapter" data-level="3.7.5" data-path="before-we-start.html"><a href="before-we-start.html#asking-for-help"><i class="fa fa-check"></i><b>3.7.5</b> Asking for help</a></li>
<li class="chapter" data-level="3.7.6" data-path="before-we-start.html"><a href="before-we-start.html#where-to-ask-for-help"><i class="fa fa-check"></i><b>3.7.6</b> Where to ask for help?</a></li>
<li class="chapter" data-level="3.7.7" data-path="before-we-start.html"><a href="before-we-start.html#more-resources"><i class="fa fa-check"></i><b>3.7.7</b> More resources</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="introduction-to-r.html"><a href="introduction-to-r.html"><i class="fa fa-check"></i><b>4</b> Introduction to R</a><ul>
<li class="chapter" data-level="4.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#creating-objects-in-r"><i class="fa fa-check"></i><b>4.1</b> Creating objects in R</a><ul>
<li class="chapter" data-level="4.1.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#comments"><i class="fa fa-check"></i><b>4.1.1</b> Comments</a></li>
<li class="chapter" data-level="4.1.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#functions-and-their-arguments"><i class="fa fa-check"></i><b>4.1.2</b> Functions and their arguments</a></li>
<li class="chapter" data-level="4.1.3" data-path="introduction-to-r.html"><a href="introduction-to-r.html#objects-vs.variables"><i class="fa fa-check"></i><b>4.1.3</b> Objects vs. variables</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#vectors-and-data-types"><i class="fa fa-check"></i><b>4.2</b> Vectors and data types</a></li>
<li class="chapter" data-level="4.3" data-path="introduction-to-r.html"><a href="introduction-to-r.html#subsetting-vectors"><i class="fa fa-check"></i><b>4.3</b> Subsetting vectors</a><ul>
<li class="chapter" data-level="4.3.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#conditional-subsetting"><i class="fa fa-check"></i><b>4.3.1</b> Conditional subsetting</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="introduction-to-r.html"><a href="introduction-to-r.html#missing-data"><i class="fa fa-check"></i><b>4.4</b> Missing data</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="starting-with-data.html"><a href="starting-with-data.html"><i class="fa fa-check"></i><b>5</b> Starting with data</a><ul>
<li class="chapter" data-level="5.1" data-path="starting-with-data.html"><a href="starting-with-data.html#presentation-of-the-survey-data"><i class="fa fa-check"></i><b>5.1</b> Presentation of the Survey Data</a></li>
<li class="chapter" data-level="5.2" data-path="starting-with-data.html"><a href="starting-with-data.html#what-are-data-frames"><i class="fa fa-check"></i><b>5.2</b> What are data frames?</a></li>
<li class="chapter" data-level="5.3" data-path="starting-with-data.html"><a href="starting-with-data.html#inspecting-data.frame-objects"><i class="fa fa-check"></i><b>5.3</b> Inspecting <code>data.frame</code> Objects</a></li>
<li class="chapter" data-level="5.4" data-path="starting-with-data.html"><a href="starting-with-data.html#indexing-and-subsetting-data-frames"><i class="fa fa-check"></i><b>5.4</b> Indexing and subsetting data frames</a></li>
<li class="chapter" data-level="5.5" data-path="starting-with-data.html"><a href="starting-with-data.html#factors"><i class="fa fa-check"></i><b>5.5</b> Factors</a></li>
<li class="chapter" data-level="5.6" data-path="starting-with-data.html"><a href="starting-with-data.html#formatting-dates"><i class="fa fa-check"></i><b>5.6</b> Formatting Dates</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="manipulating-and-analyzing-data.html"><a href="manipulating-and-analyzing-data.html"><i class="fa fa-check"></i><b>6</b> Manipulating and analyzing data</a><ul>
<li class="chapter" data-level="6.1" data-path="manipulating-and-analyzing-data.html"><a href="manipulating-and-analyzing-data.html#what-are-dplyr-and-tidyr"><i class="fa fa-check"></i><b>6.1</b> What are <strong><code>dplyr</code></strong> and <strong><code>tidyr</code></strong>?</a></li>
<li class="chapter" data-level="6.2" data-path="manipulating-and-analyzing-data.html"><a href="manipulating-and-analyzing-data.html#selecting-columns-and-filtering-rows"><i class="fa fa-check"></i><b>6.2</b> Selecting columns and filtering rows</a></li>
<li class="chapter" data-level="6.3" data-path="manipulating-and-analyzing-data.html"><a href="manipulating-and-analyzing-data.html#pipes"><i class="fa fa-check"></i><b>6.3</b> Pipes</a><ul>
<li class="chapter" data-level="6.3.1" data-path="manipulating-and-analyzing-data.html"><a href="manipulating-and-analyzing-data.html#mutate"><i class="fa fa-check"></i><b>6.3.1</b> Mutate</a></li>
<li class="chapter" data-level="6.3.2" data-path="manipulating-and-analyzing-data.html"><a href="manipulating-and-analyzing-data.html#split-apply-combine-data-analysis-and-the-summarize-function"><i class="fa fa-check"></i><b>6.3.2</b> Split-apply-combine data analysis and the summarize() function</a></li>
<li class="chapter" data-level="6.3.3" data-path="manipulating-and-analyzing-data.html"><a href="manipulating-and-analyzing-data.html#reshaping-with-gather-and-spread"><i class="fa fa-check"></i><b>6.3.3</b> Reshaping with gather and spread</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="manipulating-and-analyzing-data.html"><a href="manipulating-and-analyzing-data.html#exporting-data"><i class="fa fa-check"></i><b>6.4</b> Exporting data</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html"><i class="fa fa-check"></i><b>7</b> Data visualization with ggplot2</a><ul>
<li class="chapter" data-level="7.1" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#plotting-with-ggplot2"><i class="fa fa-check"></i><b>7.1</b> Plotting with <strong><code>ggplot2</code></strong></a></li>
<li class="chapter" data-level="7.2" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#building-your-plots-iteratively"><i class="fa fa-check"></i><b>7.2</b> Building your plots iteratively</a></li>
<li class="chapter" data-level="7.3" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#boxplot"><i class="fa fa-check"></i><b>7.3</b> Boxplot</a></li>
<li class="chapter" data-level="7.4" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#plotting-time-series-data"><i class="fa fa-check"></i><b>7.4</b> Plotting time series data</a></li>
<li class="chapter" data-level="7.5" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#faceting"><i class="fa fa-check"></i><b>7.5</b> Faceting</a></li>
<li class="chapter" data-level="7.6" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#ggplot2-themes"><i class="fa fa-check"></i><b>7.6</b> <strong><code>ggplot2</code></strong> themes</a></li>
<li class="chapter" data-level="7.7" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#customization"><i class="fa fa-check"></i><b>7.7</b> Customization</a></li>
<li class="chapter" data-level="7.8" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#arranging-and-exporting-plots"><i class="fa fa-check"></i><b>7.8</b> Arranging and exporting plots</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="reproducible-research-with-rmarkdown.html"><a href="reproducible-research-with-rmarkdown.html"><i class="fa fa-check"></i><b>8</b> Reproducible research with Rmarkdown</a></li>
<li class="chapter" data-level="9" data-path="r-package-development.html"><a href="r-package-development.html"><i class="fa fa-check"></i><b>9</b> R Package Development</a><ul>
<li class="chapter" data-level="9.1" data-path="r-package-development.html"><a href="r-package-development.html#introduction-2"><i class="fa fa-check"></i><b>9.1</b> Introduction</a><ul>
<li class="chapter" data-level="9.1.1" data-path="r-package-development.html"><a href="r-package-development.html#references"><i class="fa fa-check"></i><b>9.1.1</b> References</a></li>
<li class="chapter" data-level="9.1.2" data-path="r-package-development.html"><a href="r-package-development.html#terminology"><i class="fa fa-check"></i><b>9.1.2</b> Terminology</a></li>
<li class="chapter" data-level="9.1.3" data-path="r-package-development.html"><a href="r-package-development.html#requirement"><i class="fa fa-check"></i><b>9.1.3</b> Requirement</a></li>
<li class="chapter" data-level="9.1.4" data-path="r-package-development.html"><a href="r-package-development.html#course-content"><i class="fa fa-check"></i><b>9.1.4</b> Course content</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="r-package-development.html"><a href="r-package-development.html#preparing-r-code"><i class="fa fa-check"></i><b>9.2</b> Preparing R code</a></li>
<li class="chapter" data-level="9.3" data-path="r-package-development.html"><a href="r-package-development.html#package-layout"><i class="fa fa-check"></i><b>9.3</b> Package layout</a></li>
<li class="chapter" data-level="9.4" data-path="r-package-development.html"><a href="r-package-development.html#package-developement-cycle"><i class="fa fa-check"></i><b>9.4</b> Package developement cycle</a></li>
<li class="chapter" data-level="9.5" data-path="r-package-development.html"><a href="r-package-development.html#package-metadata"><i class="fa fa-check"></i><b>9.5</b> Package metadata</a></li>
<li class="chapter" data-level="9.6" data-path="r-package-development.html"><a href="r-package-development.html#namespace"><i class="fa fa-check"></i><b>9.6</b> <code>NAMESPACE</code></a></li>
<li class="chapter" data-level="9.7" data-path="r-package-development.html"><a href="r-package-development.html#r-code"><i class="fa fa-check"></i><b>9.7</b> R code</a></li>
<li class="chapter" data-level="9.8" data-path="r-package-development.html"><a href="r-package-development.html#package-sub-directories"><i class="fa fa-check"></i><b>9.8</b> Package sub-directories</a></li>
<li class="chapter" data-level="9.9" data-path="r-package-development.html"><a href="r-package-development.html#documentation"><i class="fa fa-check"></i><b>9.9</b> Documentation</a></li>
<li class="chapter" data-level="9.10" data-path="r-package-development.html"><a href="r-package-development.html#manual-pages"><i class="fa fa-check"></i><b>9.10</b> Manual pages</a><ul>
<li class="chapter" data-level="9.10.1" data-path="r-package-development.html"><a href="r-package-development.html#use-roxygen"><i class="fa fa-check"></i><b>9.10.1</b> Use roxygen</a></li>
<li class="chapter" data-level="9.10.2" data-path="r-package-development.html"><a href="r-package-development.html#vignettes"><i class="fa fa-check"></i><b>9.10.2</b> Vignettes</a></li>
</ul></li>
<li class="chapter" data-level="9.11" data-path="r-package-development.html"><a href="r-package-development.html#additional-files"><i class="fa fa-check"></i><b>9.11</b> Additional files</a></li>
<li class="chapter" data-level="9.12" data-path="r-package-development.html"><a href="r-package-development.html#distributing-packages"><i class="fa fa-check"></i><b>9.12</b> Distributing packages</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html"><i class="fa fa-check"></i><b>10</b> Visualisation of high-dimensional data in R</a><ul>
<li class="chapter" data-level="10.1" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#introduction-3"><i class="fa fa-check"></i><b>10.1</b> Introduction</a></li>
<li class="chapter" data-level="10.2" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#data"><i class="fa fa-check"></i><b>10.2</b> Data</a></li>
<li class="chapter" data-level="10.3" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#k-means-clustering"><i class="fa fa-check"></i><b>10.3</b> K-means clustering</a><ul>
<li class="chapter" data-level="10.3.1" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#how-does-k-means-work"><i class="fa fa-check"></i><b>10.3.1</b> How does k-means work</a></li>
<li class="chapter" data-level="10.3.2" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#model-selection"><i class="fa fa-check"></i><b>10.3.2</b> Model selection</a></li>
<li class="chapter" data-level="10.3.3" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#how-to-determine-the-number-of-clusters"><i class="fa fa-check"></i><b>10.3.3</b> How to determine the number of clusters</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#hierarchical-clustering"><i class="fa fa-check"></i><b>10.4</b> Hierarchical clustering</a><ul>
<li class="chapter" data-level="10.4.1" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#how-does-hierarchical-clustering-work"><i class="fa fa-check"></i><b>10.4.1</b> How does hierarchical clustering work</a></li>
<li class="chapter" data-level="10.4.2" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#defining-clusters"><i class="fa fa-check"></i><b>10.4.2</b> Defining clusters</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>10.5</b> Principal component analysis (PCA)</a><ul>
<li class="chapter" data-level="10.5.1" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#visualisation"><i class="fa fa-check"></i><b>10.5.1</b> Visualisation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html"><i class="fa fa-check"></i><b>11</b> Interactive visualisation</a><ul>
<li class="chapter" data-level="11.1" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactivity-graphs"><i class="fa fa-check"></i><b>11.1</b> Interactivity graphs</a><ul>
<li class="chapter" data-level="11.1.1" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactivity"><i class="fa fa-check"></i><b>11.1.1</b> Interactivity</a></li>
<li class="chapter" data-level="11.1.2" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#layers"><i class="fa fa-check"></i><b>11.1.2</b> Layers</a></li>
<li class="chapter" data-level="11.1.3" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#more-components"><i class="fa fa-check"></i><b>11.1.3</b> More components</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactive-apps"><i class="fa fa-check"></i><b>11.2</b> Interactive apps</a><ul>
<li class="chapter" data-level="11.2.1" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#introduction-4"><i class="fa fa-check"></i><b>11.2.1</b> Introduction</a></li>
<li class="chapter" data-level="11.2.2" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#creation-of-our-fist-shiny-app"><i class="fa fa-check"></i><b>11.2.2</b> Creation of our fist shiny app</a></li>
<li class="chapter" data-level="11.2.3" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#the-shiny-infrastructure"><i class="fa fa-check"></i><b>11.2.3</b> The shiny infrastructure</a></li>
<li class="chapter" data-level="11.2.4" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#another-shiny-app"><i class="fa fa-check"></i><b>11.2.4</b> Another shiny app</a></li>
<li class="chapter" data-level="11.2.5" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#single-file-app"><i class="fa fa-check"></i><b>11.2.5</b> Single-file app</a></li>
<li class="chapter" data-level="11.2.6" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#theres-more-to-shiny"><i class="fa fa-check"></i><b>11.2.6</b> There’s more to <code>shiny</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="navigating-the-bioconductor-project.html"><a href="navigating-the-bioconductor-project.html"><i class="fa fa-check"></i><b>12</b> Navigating the Bioconductor project</a><ul>
<li class="chapter" data-level="12.1" data-path="navigating-the-bioconductor-project.html"><a href="navigating-the-bioconductor-project.html#biocviews"><i class="fa fa-check"></i><b>12.1</b> <em>biocViews</em></a></li>
<li class="chapter" data-level="12.2" data-path="navigating-the-bioconductor-project.html"><a href="navigating-the-bioconductor-project.html#workflows"><i class="fa fa-check"></i><b>12.2</b> Workflows</a></li>
<li class="chapter" data-level="12.3" data-path="navigating-the-bioconductor-project.html"><a href="navigating-the-bioconductor-project.html#learning-about-specific-packages"><i class="fa fa-check"></i><b>12.3</b> Learning about specific packages</a><ul>
<li class="chapter" data-level="12.3.1" data-path="navigating-the-bioconductor-project.html"><a href="navigating-the-bioconductor-project.html#package-vignettes"><i class="fa fa-check"></i><b>12.3.1</b> Package vignettes</a></li>
<li class="chapter" data-level="12.3.2" data-path="navigating-the-bioconductor-project.html"><a href="navigating-the-bioconductor-project.html#installation-1"><i class="fa fa-check"></i><b>12.3.2</b> Installation</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="navigating-the-bioconductor-project.html"><a href="navigating-the-bioconductor-project.html#versions"><i class="fa fa-check"></i><b>12.4</b> Versions</a></li>
<li class="chapter" data-level="12.5" data-path="navigating-the-bioconductor-project.html"><a href="navigating-the-bioconductor-project.html#getting-help"><i class="fa fa-check"></i><b>12.5</b> Getting help</a></li>
<li class="chapter" data-level="12.6" data-path="navigating-the-bioconductor-project.html"><a href="navigating-the-bioconductor-project.html#data-infrastructure"><i class="fa fa-check"></i><b>12.6</b> Data infrastructure</a></li>
<li class="chapter" data-level="12.7" data-path="navigating-the-bioconductor-project.html"><a href="navigating-the-bioconductor-project.html#more-resources-1"><i class="fa fa-check"></i><b>12.7</b> More resources</a></li>
<li class="chapter" data-level="12.8" data-path="navigating-the-bioconductor-project.html"><a href="navigating-the-bioconductor-project.html#challenges"><i class="fa fa-check"></i><b>12.8</b> Challenges</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="r-programming-concepts-and-tools.html"><a href="r-programming-concepts-and-tools.html"><i class="fa fa-check"></i><b>13</b> R programming concepts and tools</a><ul>
<li class="chapter" data-level="13.1" data-path="r-programming-concepts-and-tools.html"><a href="r-programming-concepts-and-tools.html#defensive-programming"><i class="fa fa-check"></i><b>13.1</b> Defensive programming</a><ul>
<li class="chapter" data-level="" data-path="r-programming-concepts-and-tools.html"><a href="r-programming-concepts-and-tools.html#talking-to-users"><i class="fa fa-check"></i>Talking to users</a></li>
</ul></li>
<li class="chapter" data-level="13.2" data-path="r-programming-concepts-and-tools.html"><a href="r-programming-concepts-and-tools.html#kiss"><i class="fa fa-check"></i><b>13.2</b> KISS</a></li>
<li class="chapter" data-level="13.3" data-path="r-programming-concepts-and-tools.html"><a href="r-programming-concepts-and-tools.html#failing-fast-and-well"><i class="fa fa-check"></i><b>13.3</b> Failing fast and well</a></li>
<li class="chapter" data-level="13.4" data-path="r-programming-concepts-and-tools.html"><a href="r-programming-concepts-and-tools.html#consistency-and-predictability"><i class="fa fa-check"></i><b>13.4</b> Consistency and predictability</a></li>
<li class="chapter" data-level="13.5" data-path="r-programming-concepts-and-tools.html"><a href="r-programming-concepts-and-tools.html#comparisons"><i class="fa fa-check"></i><b>13.5</b> Comparisons</a><ul>
<li class="chapter" data-level="" data-path="r-programming-concepts-and-tools.html"><a href="r-programming-concepts-and-tools.html#floating-point-issues-to-be-aware-of"><i class="fa fa-check"></i>Floating point issues to be aware of</a></li>
<li class="chapter" data-level="" data-path="r-programming-concepts-and-tools.html"><a href="r-programming-concepts-and-tools.html#floating-point-how-to-compare"><i class="fa fa-check"></i>Floating point: how to compare</a></li>
<li class="chapter" data-level="" data-path="r-programming-concepts-and-tools.html"><a href="r-programming-concepts-and-tools.html#exact-identity"><i class="fa fa-check"></i>Exact identity</a></li>
</ul></li>
<li class="chapter" data-level="13.6" data-path="r-programming-concepts-and-tools.html"><a href="r-programming-concepts-and-tools.html#exercise"><i class="fa fa-check"></i><b>13.6</b> Exercise</a></li>
<li class="chapter" data-level="13.7" data-path="r-programming-concepts-and-tools.html"><a href="r-programming-concepts-and-tools.html#debugging-techniques-and-tools"><i class="fa fa-check"></i><b>13.7</b> Debugging: techniques and tools</a><ul>
<li class="chapter" data-level="" data-path="r-programming-concepts-and-tools.html"><a href="r-programming-concepts-and-tools.html#tools"><i class="fa fa-check"></i>Tools</a></li>
<li class="chapter" data-level="" data-path="r-programming-concepts-and-tools.html"><a href="r-programming-concepts-and-tools.html#exercise-1"><i class="fa fa-check"></i>Exercise</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Data analysis and R programming</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="navigating-the-bioconductor-project" class="section level1">
<h1><span class="header-section-number">Chapter 12</span> Navigating the Bioconductor project</h1>
<div id="biocviews" class="section level2">
<h2><span class="header-section-number">12.1</span> <em>biocViews</em></h2>
<p>Bioconductor has become a large project proposing many packages (1473 software packages at the time of writing) across many domains of high throughput biology. It continues to grow, at an increasing rate, and it can be difficult to get started.</p>
<p>One way to find packages of interest is to navigate the <em>biocViews</em> hierarchy. Every package is tagged with a set of <em>biocViews</em> labels. The highest level defines 3 types of packages:</p>
<ul>
<li>Software: packages providing a specific functionality.</li>
<li>AnnotationData: packages providing annotations, such as various ontologies, species annotations, microarray annotations, …</li>
<li>ExperimentData: packages distributing experiments.</li>
</ul>
<p>The <em>biocViews</em> page is available here</p>
<ul>
<li><a href="https://bioconductor.org/packages/release/BiocViews.html#___Software" class="uri">https://bioconductor.org/packages/release/BiocViews.html#___Software</a></li>
</ul>
<p>It is most easily accessed by clicking on the <em>software packages</em> link on the homepage, under <em>About Bioconductor</em>.</p>
<div class="figure">
<img src="figs/biocwww.png" />

</div>
<p>See also this <a href="https://bioconductor.org/developers/how-to/biocViews/">page</a> for additional information.</p>
</div>
<div id="workflows" class="section level2">
<h2><span class="header-section-number">12.2</span> Workflows</h2>
<p>On the other hand, people generally don’t approach the Bioconductor project to learn the whole project, but are interested by a specific analysis from a Bioconductor package, that they have read in a paper of interest. In my opinion, it is more effective to restrict ones attention to a problem or analysis of interest to first immerse oneself into Bioconductor, then broaden up ones experience to other topics and packages.</p>
<p>To to that, the project offers workflows that provide a general introduction to topics such as sequence analysis, annotation resources, RNA-Seq data analyis, Mass spectrometry and proteomics, CyTOF analysis, ….</p>
<ul>
<li><a href="https://bioconductor.org/help/workflows/" class="uri">https://bioconductor.org/help/workflows/</a></li>
</ul>
<p>A similar set of resources are published in <a href="https://f1000research.com/">F1000Research</a> under the Bioconductor gateway</p>
<ul>
<li><a href="https://f1000research.com/gateways/bioconductor" class="uri">https://f1000research.com/gateways/bioconductor</a></li>
</ul>
<p>These peer-reviewed papers describe more complete pipelines involving one or several packages.</p>
</div>
<div id="learning-about-specific-packages" class="section level2">
<h2><span class="header-section-number">12.3</span> Learning about specific packages</h2>
<p>Each Bioconductor package has it’s own <em>landing pages</em> that provides all necessary information about a package, including a short summary, its current version, the authors, how the cite the package, installation instructions, and links to all package vignettes.</p>
<p>Any Bioconductor package page can be contructed by appending the package’s name to <code>https://bioconductor.org/packages/</code> to produce an URL like</p>
<ul>
<li><a href="https://bioconductor.org/packages/packagename" class="uri">https://bioconductor.org/packages/packagename</a></li>
</ul>
<p>This works for any type of package (software, annotation or data). For example, the pages for packages <em><a href="http://bioconductor.org/packages/DESeq2">DESeq2</a></em> or <em><a href="http://bioconductor.org/packages/MSnbase">MSnbase</a></em> would be</p>
<ul>
<li><a href="https://bioconductor.org/packages/DESeq2" class="uri">https://bioconductor.org/packages/DESeq2</a></li>
</ul>
<p>and</p>
<ul>
<li><a href="https://bioconductor.org/packages/MSnbase" class="uri">https://bioconductor.org/packages/MSnbase</a></li>
</ul>
<p>These short URLs are then resolved to their longer form to redirect to the longer package URL leading the user to the current release version of the packge.</p>
<div id="package-vignettes" class="section level3">
<h3><span class="header-section-number">12.3.1</span> Package vignettes</h3>
<p>An important quality of every Bioconductor package is the availability of a dedicated <em>vignette</em>. Vignettes are documentations (generally provided as pdf or html files) that provide a generic overview of the package, without necessarily going in detail for every function of the package.</p>
<p>Vignettes are special in that respect as they are produced as part of the package building process. The code in a vignette is executed and its output, whether in the form of simple text, tables and figures, are inserted in the vignette before the final file (in pdf or html) is produced. Hence, all the code and outputs are guaranteed to be correct and reproduced.</p>
<p>Given a vignette, it is this possible to re-generate all the results. To make reproducing a long vignette as easy as possible without copy and pasting all code chunks one by one, it is possible to extract the code into an R script runnung the <code>Stangle</code> (from the <code>utils</code> package - see <a href="https://bioconductor.org/help/package-vignettes/">here</a> for details) or <code>knitr::purl</code> functions on the vignette source document.</p>
</div>
<div id="installation-1" class="section level3">
<h3><span class="header-section-number">12.3.2</span> Installation</h3>
<p>The installation of a Bioconductor package is done using the <code>biocLite</code> function from the <code>BiocInstaller</code> package. The first time, when the package isn’t available yet, you need to run</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## try http:// if https:// URLs are not supported
<span class="kw">source</span>(<span class="st">&quot;https://bioconductor.org/biocLite.R&quot;</span>)</code></pre></div>
<p>The above command will install and load <code>r Biocpkg(&quot;BiocInstaller&quot;)</code>. If it isn’t loaded yet, first load it with <code>library</code>, then install your package(s) of interest with <code>biocLite</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;BiocInstaller&quot;</span>)
<span class="kw">biocLite</span>(<span class="st">&quot;Biobase&quot;</span>)</code></pre></div>
<p>This <a href="https://bioconductor.org/install/">page</a> provides more details on package installation and update.</p>
</div>
</div>
<div id="versions" class="section level2">
<h2><span class="header-section-number">12.4</span> Versions</h2>
<p>It is also useful to know that at any given time, there are two Bioconductor versions - there is always a release (stable) and a development (devel) versions. For example, in October 2017, the release version is 3.6 and the development is 3.7.</p>
<p>The individual packages have a similar scheme. Every package is available for the release and the development versions of Bioconductor. These two versions of the package have also different version numbers, where the last digit is even for the former and off for the later. Currently, the <code>rBiocpkg(&quot;MSnbase&quot;)</code> has versions <code>2.4.0</code> and <code>2.5.1</code>, respectively.</p>
<p>Finally, every Bioconductor version is tight to an R version. To access the current Bioconductor release 3.6 version, one needs to use the latest R version, which is 3.4.2. Hence, it is important to have an up-to-date R installation to keep up with the latest developments in Bioconductor. More details <a href="https://bioconductor.org/developers/how-to/version-numbering/">here</a>.</p>
</div>
<div id="getting-help" class="section level2">
<h2><span class="header-section-number">12.5</span> Getting help</h2>
<p>The best way to get help with regard the a Bioconductor package is to post the question on the <em>Bioconductor support forum</em> at <a href="https://support.bioconductor.org/" class="uri">https://support.bioconductor.org/</a>. Package developers generally follow the support site for questions related to their packages. See this page for <a href="https://bioconductor.org/help/support/">some details</a>.</p>
<p>To maximise the chances to obtain a answer promptly, it is important to provide details for other to understand the question and, if relevant, reproduce the observed errors. The Bioconductor project has a dedicated <a href="https://bioconductor.org/help/support/posting-guide/">posting guide</a>. Here’s another useful guide on <a href="http://adv-r.had.co.nz/Reproducibility.html">how to write a reproducible question</a>.</p>
<p>Packages come with a lot of documentation build in, that users are advised to read to familiarise themselves with the package and how to use it. In addition to the package vignettes are describe above, every function of class in a package is documented in great detail in their respective <em>man</em> page, that can be accessed with <code>?function</code>.</p>
<p>There is also a dedicated <a href="https://bioconductor.org/help/support/posting-guide/"><em>developer mailing list</em></a> that is dedicated for questions and discussions related to package development.</p>
</div>
<div id="data-infrastructure" class="section level2">
<h2><span class="header-section-number">12.6</span> Data infrastructure</h2>
<p>An essential aspect that is central to Bioconductor and its success is the availability of core data infrastructure that is used across packages. Package developers are advised to make use of existing infrastructure to provide coherence, interoperability and stability to the project as a whole.</p>
<p>Here are some core classes, taken from the <a href="https://bioconductor.org/developers/how-to/commonMethodsAndClasses/">Common Bioconductor Methods and Classes</a> page:</p>
<div id="importing" class="section level4 unnumbered">
<h4>Importing</h4>
<ul>
<li>GTF, GFF, BED, BigWig, etc., - <em><a href="http://bioconductor.org/packages/rtracklayer">rtracklayer</a></em><code>::import()</code></li>
<li>VCF – <em><a href="http://bioconductor.org/packages/VariantAnnotation">VariantAnnotation</a></em><code>::readVcf()</code></li>
<li>SAM / BAM – <em><a href="http://bioconductor.org/packages/Rsamtools">Rsamtools</a></em><code>::scanBam()</code>, <em><a href="http://bioconductor.org/packages/GenomicAlignments">GenomicAlignments</a></em><code>:readGAlignment*()</code></li>
<li>FASTA – <em><a href="http://bioconductor.org/packages/Biostrings">Biostrings</a></em><code>::readDNAStringSet()</code></li>
<li>FASTQ – <em><a href="http://bioconductor.org/packages/ShortRead">ShortRead</a></em><code>::readFastq()</code></li>
<li>Mass spectrometry data (XML-based and mfg formats) – <em><a href="http://bioconductor.org/packages/MSnbase">MSnbase</a></em><code>::readMSData()</code>, <em><a href="http://bioconductor.org/packages/MSnbase">MSnbase</a></em><code>::readMgfData()</code></li>
</ul>
</div>
<div id="common-classes" class="section level4 unnumbered">
<h4>Common Classes</h4>
<ul>
<li>Rectangular feature x sample data – <em><a href="http://bioconductor.org/packages/SummarizedExperiment">SummarizedExperiment</a></em><code>::SummarizedExperiment()</code> (RNAseq count matrix, microarray, …)</li>
<li>Genomic coordinates – <em><a href="http://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></em><code>::GRanges()</code> (1-based, closed interval)</li>
<li>DNA / RNA / AA sequences – <em><a href="http://bioconductor.org/packages/Biostrings">Biostrings</a></em><code>::*StringSet()</code></li>
<li>Gene sets – <em><a href="http://bioconductor.org/packages/GSEABase">GSEABase</a></em><code>::GeneSet()</code> <em><a href="http://bioconductor.org/packages/GSEABase">GSEABase</a></em><code>::GeneSetCollection()</code></li>
<li>Multi-omics data – <em><a href="http://bioconductor.org/packages/MultiAssayExperiment">MultiAssayExperiment</a></em><code>::MultiAssayExperiment()</code></li>
<li>Single cell data – <em><a href="http://bioconductor.org/packages/SingleCellExperiment">SingleCellExperiment</a></em><code>::SingleCellExperiment()</code></li>
<li>Mass spectrometry data – <em><a href="http://bioconductor.org/packages/MSnbase">MSnbase</a></em><code>::MSnExp()</code></li>
</ul>
<p>A common design theme that is found throughout many Bioconductor core classes is illustrated below, which is found in microarrays, quantiative proteomics data, RNA-Seq data, … It contains a rectangular feature x sample data matrix and sample and feature-specific annotations.</p>
<div class="figure">
<img src="figs/msnset.png" />

</div>
</div>
</div>
<div id="more-resources-1" class="section level2">
<h2><span class="header-section-number">12.7</span> More resources</h2>
<ul>
<li>Videos: <a href="https://www.youtube.com/user/bioconductor" class="uri">https://www.youtube.com/user/bioconductor</a></li>
<li>Community Contributed Help Resources: <a href="https://bioconductor.org/help/community/" class="uri">https://bioconductor.org/help/community/</a></li>
<li>Tutorials: <a href="https://support.bioconductor.org/t/Tutorials/" class="uri">https://support.bioconductor.org/t/Tutorials/</a></li>
</ul>
</div>
<div id="challenges" class="section level2">
<h2><span class="header-section-number">12.8</span> Challenges</h2>
<ol style="list-style-type: decimal">
<li><p>Install a Bioconductor package of your choice, discover the vignette(s) it offers, open one, and extract the R code of it.</p></li>
<li><p>Find a package that allows reading raw mass spectrometry data and identify the specific function. Either use the <em>biocViews</em> tree, look for a possible <em>workflow</em>, or look in the common methods and classes page on the Bioconductor page.</p></li>
</ol>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="interactive-visualisation.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="r-programming-concepts-and-tools.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
