<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Data analysis and R programming</title>
  <meta name="description" content="Material for the ‘Data analysis and R programming’ course.">
  <meta name="generator" content="bookdown 0.5 and GitBook 2.6.7">

  <meta property="og:title" content="Data analysis and R programming" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Material for the ‘Data analysis and R programming’ course." />
  <meta name="github-repo" content="lgatto/2017_11_09_Rcourse_Jena" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Data analysis and R programming" />
  
  <meta name="twitter:description" content="Material for the ‘Data analysis and R programming’ course." />
  

<meta name="author" content="Laurent Gatto">


<meta name="date" content="2017-11-02">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="introduction.html">
<link rel="next" href="visualisation-of-high-dimensional-data-in-r.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/htmlwidgets-0.9/htmlwidgets.js"></script>
<script src="libs/datatables-binding-0.2/datatables.js"></script>
<link href="libs/dt-core-1.10.12/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.10.12/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.10.12/js/jquery.dataTables.min.js"></script>


<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.1</b> Installation</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.2</b> License</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>2</b> Introduction</a><ul>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#programme"><i class="fa fa-check"></i>Programme</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="r-package-development.html"><a href="r-package-development.html"><i class="fa fa-check"></i><b>3</b> R Package Development</a><ul>
<li class="chapter" data-level="3.1" data-path="r-package-development.html"><a href="r-package-development.html#introduction-1"><i class="fa fa-check"></i><b>3.1</b> Introduction</a><ul>
<li class="chapter" data-level="3.1.1" data-path="r-package-development.html"><a href="r-package-development.html#references"><i class="fa fa-check"></i><b>3.1.1</b> References</a></li>
<li class="chapter" data-level="3.1.2" data-path="r-package-development.html"><a href="r-package-development.html#terminology"><i class="fa fa-check"></i><b>3.1.2</b> Terminology</a></li>
<li class="chapter" data-level="3.1.3" data-path="r-package-development.html"><a href="r-package-development.html#requirement"><i class="fa fa-check"></i><b>3.1.3</b> Requirement</a></li>
<li class="chapter" data-level="3.1.4" data-path="r-package-development.html"><a href="r-package-development.html#course-content"><i class="fa fa-check"></i><b>3.1.4</b> Course content</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="r-package-development.html"><a href="r-package-development.html#preparing-r-code"><i class="fa fa-check"></i><b>3.2</b> Preparing R code</a></li>
<li class="chapter" data-level="3.3" data-path="r-package-development.html"><a href="r-package-development.html#package-layout"><i class="fa fa-check"></i><b>3.3</b> Package layout</a></li>
<li class="chapter" data-level="3.4" data-path="r-package-development.html"><a href="r-package-development.html#package-developement-cycle"><i class="fa fa-check"></i><b>3.4</b> Package developement cycle</a></li>
<li class="chapter" data-level="3.5" data-path="r-package-development.html"><a href="r-package-development.html#package-metadata"><i class="fa fa-check"></i><b>3.5</b> Package metadata</a></li>
<li class="chapter" data-level="3.6" data-path="r-package-development.html"><a href="r-package-development.html#namespace"><i class="fa fa-check"></i><b>3.6</b> <code>NAMESPACE</code></a></li>
<li class="chapter" data-level="3.7" data-path="r-package-development.html"><a href="r-package-development.html#r-code"><i class="fa fa-check"></i><b>3.7</b> R code</a></li>
<li class="chapter" data-level="3.8" data-path="r-package-development.html"><a href="r-package-development.html#package-sub-directories"><i class="fa fa-check"></i><b>3.8</b> Package sub-directories</a></li>
<li class="chapter" data-level="3.9" data-path="r-package-development.html"><a href="r-package-development.html#documentation"><i class="fa fa-check"></i><b>3.9</b> Documentation</a></li>
<li class="chapter" data-level="3.10" data-path="r-package-development.html"><a href="r-package-development.html#manual-pages"><i class="fa fa-check"></i><b>3.10</b> Manual pages</a><ul>
<li class="chapter" data-level="3.10.1" data-path="r-package-development.html"><a href="r-package-development.html#use-roxygen"><i class="fa fa-check"></i><b>3.10.1</b> Use roxygen</a></li>
<li class="chapter" data-level="3.10.2" data-path="r-package-development.html"><a href="r-package-development.html#vignettes"><i class="fa fa-check"></i><b>3.10.2</b> Vignettes</a></li>
</ul></li>
<li class="chapter" data-level="3.11" data-path="r-package-development.html"><a href="r-package-development.html#additional-files"><i class="fa fa-check"></i><b>3.11</b> Additional files</a></li>
<li class="chapter" data-level="3.12" data-path="r-package-development.html"><a href="r-package-development.html#distributing-packages"><i class="fa fa-check"></i><b>3.12</b> Distributing packages</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html"><i class="fa fa-check"></i><b>4</b> Visualisation of high-dimensional data in R</a><ul>
<li class="chapter" data-level="4.1" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#introduction-2"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#data"><i class="fa fa-check"></i><b>4.2</b> Data</a></li>
<li class="chapter" data-level="4.3" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#k-means-clustering"><i class="fa fa-check"></i><b>4.3</b> K-means clustering</a><ul>
<li class="chapter" data-level="4.3.1" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#how-does-k-means-work"><i class="fa fa-check"></i><b>4.3.1</b> How does k-means work</a></li>
<li class="chapter" data-level="4.3.2" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#model-selection"><i class="fa fa-check"></i><b>4.3.2</b> Model selection</a></li>
<li class="chapter" data-level="4.3.3" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#how-to-determine-the-number-of-clusters"><i class="fa fa-check"></i><b>4.3.3</b> How to determine the number of clusters</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#hierarchical-clustering"><i class="fa fa-check"></i><b>4.4</b> Hierarchical clustering</a><ul>
<li class="chapter" data-level="4.4.1" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#how-does-hierarchical-clustering-work"><i class="fa fa-check"></i><b>4.4.1</b> How does hierarchical clustering work</a></li>
<li class="chapter" data-level="4.4.2" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#defining-clusters"><i class="fa fa-check"></i><b>4.4.2</b> Defining clusters</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>4.5</b> Principal component analysis (PCA)</a><ul>
<li class="chapter" data-level="4.5.1" data-path="visualisation-of-high-dimensional-data-in-r.html"><a href="visualisation-of-high-dimensional-data-in-r.html#visualisation"><i class="fa fa-check"></i><b>4.5.1</b> Visualisation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html"><i class="fa fa-check"></i><b>5</b> Interactive visualisation</a><ul>
<li class="chapter" data-level="5.1" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactivity-graphs"><i class="fa fa-check"></i><b>5.1</b> Interactivity graphs</a><ul>
<li class="chapter" data-level="5.1.1" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactivity"><i class="fa fa-check"></i><b>5.1.1</b> Interactivity</a></li>
<li class="chapter" data-level="5.1.2" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#layers"><i class="fa fa-check"></i><b>5.1.2</b> Layers</a></li>
<li class="chapter" data-level="5.1.3" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#more-components"><i class="fa fa-check"></i><b>5.1.3</b> More components</a></li>
<li class="chapter" data-level="5.1.4" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#challenge"><i class="fa fa-check"></i><b>5.1.4</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactive-apps"><i class="fa fa-check"></i><b>5.2</b> Interactive apps</a><ul>
<li class="chapter" data-level="5.2.1" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#creation-of-our-fist-shiny-app"><i class="fa fa-check"></i><b>5.2.1</b> Creation of our fist shiny app</a></li>
<li class="chapter" data-level="5.2.2" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#the-shiny-infrastructure"><i class="fa fa-check"></i><b>5.2.2</b> The shiny infrastructure</a></li>
<li class="chapter" data-level="5.2.3" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#another-shiny-app"><i class="fa fa-check"></i><b>5.2.3</b> Another shiny app</a></li>
<li class="chapter" data-level="5.2.4" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#challenges"><i class="fa fa-check"></i><b>5.2.4</b> Challenges</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Data analysis and R programming</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="r-package-development" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> R Package Development</h1>
<blockquote>
<p>This section is based on material from the <a href="R%20package%20development">https://github.com/lgatto/RPackageDevelopment</a> course.</p>
</blockquote>
<div id="introduction-1" class="section level2">
<h2><span class="header-section-number">3.1</span> Introduction</h2>
<p>Packages are the way to share R code in a structures, reproducible and tractable way. Even if the intend is not to disseminate your code, packaging it is worth it. Packages provide a mechanism for loading optional code and attached documentation as needed.</p>
<ul>
<li>logically group your own functions</li>
<li>keep code and documentation together and consistent</li>
<li>keep code and data together</li>
<li>keep track of changes in code</li>
<li>summarise all packages used for a analysis (see <code>sessionInfo()</code>)</li>
<li>make a reproducible research compendium (container for code, text, data as a means for distributing, managing and updating)</li>
<li>optionally test your code</li>
<li>… project managment</li>
</ul>
<div id="references" class="section level3">
<h3><span class="header-section-number">3.1.1</span> References</h3>
<ul>
<li><a href="http://r-pkgs.had.co.nz/">R packages</a>, by Hadley Wickham</li>
<li>R Installation and Administration [R-admin], R Core team</li>
<li>Writing R Extensions [R-ext], R Core team</li>
</ul>
<blockquote>
<p>Use <code>help.start()</code> to access them from your local installation, or <a href="http://cran.r-project.org/manuals.html" class="uri">http://cran.r-project.org/manuals.html</a> from the web.</p>
</blockquote>
</div>
<div id="terminology" class="section level3">
<h3><span class="header-section-number">3.1.2</span> Terminology</h3>
<p>A <strong>package</strong> is loaded from a <strong>library</strong> by the function <code>library()</code>. Thus a library is a directory containing installed packages.</p>
<blockquote>
<p>Calling <code>library(&quot;foo&quot;, lib.loc = &quot;/path/to/bar&quot;)</code> loads the package (book) <code>foo</code> from the library <code>bar</code> located at <code>/path/to/bar</code>.</p>
</blockquote>
</div>
<div id="requirement" class="section level3">
<h3><span class="header-section-number">3.1.3</span> Requirement</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;devtools&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;roxygen2&quot;</span>)</code></pre></div>
</div>
<div id="course-content" class="section level3">
<h3><span class="header-section-number">3.1.4</span> Course content</h3>
<p>Basic workflow</p>
<ol style="list-style-type: decimal">
<li>Prepare R code</li>
<li>Create package directory: <code>mypackage</code></li>
<li><strong>Build</strong> the package tarball</li>
<li><strong>Check</strong> the package</li>
<li><strong>Install</strong> the package</li>
</ol>
<p>Step 2 is done only once. Package developement cycles through 3 - 5.</p>
<p>Also - Writing package documentation - Vignettes - Testing packages - Compiled code</p>
</div>
</div>
<div id="preparing-r-code" class="section level2">
<h2><span class="header-section-number">3.2</span> Preparing R code</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fn &lt;-<span class="st"> </span><span class="cf">function</span>() 
    <span class="kw">message</span>(<span class="st">&quot;I love R packages&quot;</span>)    </code></pre></div>
</div>
<div id="package-layout" class="section level2">
<h2><span class="header-section-number">3.3</span> Package layout</h2>
<p>We can use</p>
<ul>
<li><code>package.skeleton(&quot;myRpackage&quot;, list = &quot;fn&quot;)</code></li>
<li><code>devtools::create(&quot;myRpackage&quot;)</code> also create an <code>.Rproj</code> file.</li>
<li>Use the RStudio wizard: <code>New Project &gt; New Directory &gt; R Package</code></li>
</ul>
<pre><code>myRpackage/
|-- DESCRIPTION
|-- NAMESPACE
|-- man
|   `-- fun.Rd
`-- R
    `-- fun.R</code></pre>
<p>This is the <em>source package</em>. From this, we need to create the <em>package tarball</em> (or <em>package bundle</em>), i.e. a compressed archive of the source. We can also create <em>binary packages</em> for Windows and Mac.</p>
</div>
<div id="package-developement-cycle" class="section level2">
<h2><span class="header-section-number">3.4</span> Package developement cycle</h2>
<p>In the shell</p>
<pre><code>R CMD build myPackage               ## creates myRpackage_1.0.tar.gz
R CMD check myPackage_1.0.tar.gz    ## create myRpackage.Rcheck
R CMD INSTALL myRpackage_1.0.tar.gz ## Installation in the default library</code></pre>
<p>Using RStudio useful keyboard shortcuts for package authoring:</p>
<ul>
<li>Build and Reload Package: <code>Ctrl + Shift + B</code></li>
<li>Check Package: <code>Ctrl + Shift + E</code></li>
<li>Test Package: <code>Ctrl + Shift + T</code></li>
</ul>
<p>Using devtools: * <code>devtools::build()</code> * <code>devtools::build(binary = TRUE)</code> * <code>devtools::check()</code> * <code>devtools::install()</code></p>
<p>A shortcut when developing:</p>
<ul>
<li><code>devtools::load_all()</code></li>
</ul>
</div>
<div id="package-metadata" class="section level2">
<h2><span class="header-section-number">3.5</span> Package metadata</h2>
<p>The <code>DESCRIPTION</code> file</p>
<pre><code>Package: myRpackage ## mandatory (*)
Type: Package ## optional, &#39;Package&#39; is default type 
Title: What the package does (short line) ## *
Version: 1.0 ##  *
Date: 2013-05-10 ## release date of the current version 
Author: Who wrote it ## *
Maintainer: Who to complain to &lt;yourfault@somewhere.net&gt; ## *
Description: More about what it does (maybe more than one line) ## *
License: What license is it under? ## *
Depends: methods, Biostrings ## for e.g.
Imports: evd ## for e.g.
Suggests: BSgenome.Hsapiens.UCSC.hg19 ## for e.g.
Collate: &#39;DataClasses.R&#39; &#39;read.R&#39; ## for e.g.</code></pre>
<p>Package dependencies:</p>
<ul>
<li><strong>Depends</strong> A comma-separated list of package names (optionally with versions) which this package depends on.</li>
<li><strong>Suggests</strong> Packages that are not necessarily needed: used only in examples, tests or vignettes, loaded in the body of functions</li>
<li><strong>Imports</strong> Packages whose name spaces are imported from (as specified in the <code>NAMESPACE</code> file) which do not need to be attached to the search path.</li>
<li><strong>Collate</strong> Controls the collation order for the R code files in a package. If filed is present, all source files must be listed.</li>
</ul>
<p>Packages are attached to the search path with  or .</p>
<ul>
<li><p><strong>Attach</strong> When a package is attached, then all of its dependencies (see <code>Depends</code> field in its <code>DESCRIPTION</code> file) are also attached. Such packages are part of the evaluation environment and will be searched.</p></li>
<li><p><strong>Load</strong> One can also use the <code>Imports</code> field in the <code>NAMESPACE</code> file. Imported packages are loaded but are not attached: they do not appear on the search path and are available only to the package that imported them.</p></li>
</ul>
</div>
<div id="namespace" class="section level2">
<h2><span class="header-section-number">3.6</span> <code>NAMESPACE</code></h2>
<p>Restricts the symbols that are exported and imports functionality from other packages. Only the exported symbols will have to be documented.</p>
<pre><code>export(f, g) ## exports f and g 
exportPattern(&quot;^[^\\.]&quot;)
import(foo) ## imports all symbols from package foo
importFrom(foo, f, g) ## imports f and g from foo</code></pre>
<p>It is possible to explicitely use symbol <code>s</code> from package <code>foo</code> with <code>foo::s</code> or <code>foo:::s</code> if <code>s</code> is not exported.</p>
</div>
<div id="r-code" class="section level2">
<h2><span class="header-section-number">3.7</span> R code</h2>
<p>Contains <code>source()</code>able R source code to be installed. Files must start with an ASCII (lower or upper case) letter or digit and have one of the extensions <code>.R</code>, <code>.S</code>, <code>.q</code>, <code>.r</code>, or <code>.s</code> (use <code>.R</code> or <code>.r</code>).</p>
<ul>
<li>General style guidelines and best practice apply.</li>
<li>Any number of files in <code>R</code>.</li>
<li>Any number of functions (methods, classes) in each source file.</li>
<li>Order matters (somehow), as the files will be sourced in the alphanumeric order. If that doesn’t fit, use the <code>collate</code> field in the <code>DESCRIPTION</code> files.</li>
</ul>
<p>Example</p>
<pre><code>## works fine without Collate field
AllGenerics.R       DataClasses.R
methods-ClassA.R    methods-ClassB.R
functions-ClassA.R  ...</code></pre>
<p><code>zzz.R</code> is generally used to define special functions used to initialize (called after a package is loaded and attached) and clean up (just before the package is detached). See <code>help(&quot;.onLoad&quot;))</code>, <code>?.First.Lib</code> and <code>?.Last.Lib</code> for more details.</p>
</div>
<div id="package-sub-directories" class="section level2">
<h2><span class="header-section-number">3.8</span> Package sub-directories</h2>
<ul>
<li><strong>vignettes</strong> directory for vignettes in Sweave or R markdown format.</li>
<li><strong>data</strong> for R code, compressed tables (<code>.tab</code>, <code>.txt</code>, or .<code>csv</code>, see <code>?data</code> for the file formats) and binary R objects. Available with <code>data()</code>.</li>
<li><strong>inst/docs</strong> for additional documentation. That’s also where the vignettes will be installed after compilation.</li>
<li><strong>inst/extdata</strong> directory for other data files, not belonging in <code>data</code>.</li>
<li><strong>tests</strong> code for unit tests (see <a href="https://github.com/lgatto/R-debugging">here</a> and <a href="https://github.com/lgatto/2016-02-25-adv-programming-EMBL">here</a>).</li>
<li><strong>src</strong> for compiled code (see the <a href="https://github.com/lgatto/rccpp/">rccpp</a> material)</li>
<li><strong>demo</strong> for demo code (see <code>?demo</code>)</li>
</ul>
</div>
<div id="documentation" class="section level2">
<h2><span class="header-section-number">3.9</span> Documentation</h2>
</div>
<div id="manual-pages" class="section level2">
<h2><span class="header-section-number">3.10</span> Manual pages</h2>
<p>Package functions, datasets, methods and classes are documented in <code>Rd</code>, a LaTeX-like format.</p>
<pre><code>% File src/library/base/man/load.Rd
% Part of the R package, http://www.R-project.org
% Copyright 1995-2014 R Core Team
% Distributed under GPL 2 or later

\name{load}
\alias{load}
\title{Reload Saved Datasets}
\description{
  Reload datasets written with the function \code{save}.
}
\usage{
load(file, envir = parent.frame(), verbose = FALSE)
}
\arguments{
  \item{file}{a (readable binary-mode) \link{connection} or a character string
    giving the name of the file to load (when \link{tilde expansion}
    is done).}
  \item{envir}{the environment where the data should be loaded.}
  \item{verbose}{should item names be printed during loading?}
}
\details{
  \code{load} can load \R objects saved in the current or any earlier
  format.  It can read a compressed file (see \code{\link{save}})
  directly from a file or from a suitable connection (including a call
  to \code{\link{url}}).

[...]

\value{
  A character vector of the names of objects created, invisibly.
}
\section{Warning}{
  Saved \R objects are binary files, even those saved with
  \code{ascii = TRUE}, so ensure that they are transferred without
  conversion of end of line markers.  \code{load} tries to detect such a
  conversion and gives an informative error message.

[...]

\examples{
## save all data
xx &lt;- pi # to ensure there is some data
save(list = ls(all = TRUE), file= &quot;all.RData&quot;)
rm(xx)

## restore the saved values to the current environment
local({
   load(&quot;all.RData&quot;)
   ls()
})

xx &lt;- exp(1:3)
## restore the saved values to the user&#39;s workspace
load(&quot;all.RData&quot;) ## which is here *equivalent* to
## load(&quot;all.RData&quot;, .GlobalEnv)
## This however annihilates all objects in .GlobalEnv with the same names !
xx # no longer exp(1:3)
rm(xx)
attach(&quot;all.RData&quot;) # safer and will warn about masked objects w/ same name in .GlobalEnv
ls(pos = 2)
##  also typically need to cleanup the search path:
detach(&quot;file:all.RData&quot;)

## clean up (the example):
unlink(&quot;all.RData&quot;)

\dontrun{
con &lt;- url(&quot;http://some.where.net/R/data/example.rda&quot;)
## print the value to see what objects were created.
print(load(con))
close(con) # url() always opens the connection
}}
\keyword{file}</code></pre>
<p>These R documentation files can then be converted into text, pdf or html:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">help</span>(<span class="st">&quot;load&quot;</span>)
<span class="kw">help</span>(<span class="st">&quot;load&quot;</span>, <span class="dt">help_type =</span> <span class="st">&quot;html&quot;</span>)
<span class="kw">help</span>(<span class="st">&quot;load&quot;</span>, <span class="dt">help_type =</span> <span class="st">&quot;pdf&quot;</span>)</code></pre></div>
<p>One can use <code>prompt</code>, <code>promptClass</code>, <code>promptMethods</code>, <code>promptPackage</code>, <code>promptPackage</code>, <code>promptData</code> to generate <code>Rd</code> templates for functions, classes, methods, packages and data.</p>
<div id="use-roxygen" class="section level3">
<h3><span class="header-section-number">3.10.1</span> Use roxygen</h3>
<p>The way documentation is managed in R packages separates the code from the documentation, which makes it easier to adapt the latter when to code is updated. In comes <code>roxygen2</code>, that allows developer to write their documentation on top of their functions:</p>
<pre><code>#&#39; Reads sequences data in fasta and create \code{DnaSeq}
#&#39; and \code{RnaSeq} instances. 
#&#39;
#&#39; This funtion reads DNA and RNA fasta files and generates
#&#39; valid \code{&quot;DnaSeq&quot;} and \code{&quot;RnaSeq&quot;} instances.
#&#39; 
#&#39; @title Read fasta files.
#&#39; @param infile  the name of the fasta file which the data are to be read from.
#&#39; @return an instance of \code{DnaSeq} or \code{RnaSeq}.
#&#39; @seealso \code{\linkS4class{GenericSeq}}, \code{\linkS4class{DnaSeq}} and \code{\linkS4class{RnaSeq}}.
#&#39; @examples
#&#39; f &lt;- dir(system.file(&quot;extdata&quot;,package=&quot;sequences&quot;),pattern=&quot;fasta&quot;,full.names=TRUE)
#&#39; f
#&#39; aa &lt;- readFasta(f)
#&#39; aa
#&#39; @author Laurent Gatto \email{lg390@@cam.ac.uk}
#&#39; @export
readFasta &lt;- function(infile){
  lines &lt;- readLines(infile)
  header &lt;- grep(&quot;^&gt;&quot;, lines)
  if (length(header)&gt;1) {
    warning(&quot;Reading first sequence only.&quot;)
    lines &lt;- lines[header[1]:(header[2]-1)]
    header &lt;- header[1]
  }
  ##### (code cut for space reasons) #####
  if (validObject(newseq))
    return(newseq)
}</code></pre>
<p>The roxygen code can then be parsed and converted to <code>Rd</code> using the <code>roxygen2::roxygenise</code> or <code>devtools::document</code> functions.</p>
<p>Note that the <code>roxygenise</code> function does more than produce documentation (that part is handled by the <code>rd</code> roclet, set with <code>roclet = &quot;rd&quot;</code>). It can also manage your <code>NAMESPACE</code> file and <code>Collate</code> field.</p>
<p>Note: recently, support for markdown format has been added to roxygen.</p>
</div>
<div id="vignettes" class="section level3">
<h3><span class="header-section-number">3.10.2</span> Vignettes</h3>
<p>While manual pages are meant to be specific and technical, vignettes are workflow-type documentation files that provide an overview and/or a use-case demonstrating the package’s functionality.</p>
<p>Vignettes can be written in Sweave format (<code>.Rnw</code> extension), supporting R code chunks in LaTeX documents, or R markdown formart (<code>.Rmd</code> extension) for R code and markdown.</p>
<p>The source document (<code>.Rnw</code> or <code>.Rmd</code>) can be</p>
<ul>
<li>weaved into <code>tex</code> or <code>md</code> files respectively</li>
<li>and converted into <code>pdf</code> or <code>html</code> (<code>Rnw</code> to <code>pdf</code> only, <code>Rmd</code> to either)</li>
</ul>
<p>using the <code>utils::Sweave</code> (<code>Rnw</code> only) or <code>knitr::knit</code> functions.</p>
<p>Rstudio makes it very easy to write and compile <code>Rmd</code> documents (independently of R packages). When inside a package, the documents are stored in the <code>vignettes</code> directory and compiled/converted automatically when the package is built.</p>
<p>Note: if you use <code>knitr</code> and <code>rmarkdown</code> for your vignette, you’ll have to add these dependencies in the <code>Suggests</code> field and specify <code>VignetteBuilder: knitr</code> in the <code>DESCRIPTION</code> file.</p>
</div>
</div>
<div id="additional-files" class="section level2">
<h2><span class="header-section-number">3.11</span> Additional files</h2>
<ul>
<li><code>.Rbuildignore</code> with a list of files/dirs to ignore when building. For example the <code>.Rproj</code> file.</li>
<li><code>.Rinstignore</code>with a list of files/dirs to ignore when installing.</li>
<li><code>CITATION</code> file (see <code>citation()</code> function)</li>
<li><code>README.Rmd</code>/<code>README.md</code> files if you use github.</li>
</ul>
</div>
<div id="distributing-packages" class="section level2">
<h2><span class="header-section-number">3.12</span> Distributing packages</h2>
<ul>
<li><p><strong>CRAN</strong> Read the CRAN Repository Policy (<a href="http://cran.r-project.org/web/packages/policies.html" class="uri">http://cran.r-project.org/web/packages/policies.html</a>). Upload your <code>--as-cran</code> checked} <code>myPackage\_x.y.z.tar.gz</code> to <code>ftp://cran.R-project.org/incoming</code> or using <code>http://CRAN.R-project.org/submit.html</code>. Your package will be installable with <code>install.packages(&quot;myRpackage&quot;)</code>.</p></li>
<li><p><strong>R-forge</strong> Log in, register a project and wait for acceptance. Then commit you code to the svn repository. Your package will be installable with <code>install.packages</code> using <code>repos=&quot;http://R-Forge.R-project.org&quot;</code>.</p></li>
<li><p><strong>GitHub</strong> (and <strong>Bitbucket</strong>) Great for development and promoting interaction and contributions. Unofficial. Autmatic checking possible through CI such as <code>travis-ci</code> for example. Packages can be installed with <code>devtools::install_github</code> (<code>devtools::install_bitbucket</code>).</p></li>
<li><p><strong>Bioconductor</strong> Make sure to satisfy submission criteria (pass <code>check</code> (and <code>BiocCheck</code>), have a vignette, use S4 if OO, make use of appropriate existing infrastructure, include a NEWS file, must <strong>not</strong> already be on CRAN, …). Your package will then be reviewed on github publicly before acceptance. A svn (git very soon) account will then be created. Package will be installable with <code>biocLite(&quot;myPackage&quot;)</code>.</p></li>
</ul>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="visualisation-of-high-dimensional-data-in-r.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
